File: sdks/typescript/src/resources/expenses.ts
import { AxiosInstance } from 'axios';
import { PaginatedResponse } from '../types';

export interface Expense {
    id: string;
    category: string;
    amount: number;
    description: string;
    date: string;
    created_at: string;
    updated_at: string;
}

export interface ExpenseCategory {
    id: string;
    name: string;
    description?: string;
}

export class Expenses {
    constructor(private client: AxiosInstance) { }

    async list(page: number = 1, pageSize: number = 20, category?: string): Promise<PaginatedResponse<Expense>> {
        const params: any = { page, page_size: pageSize };
        if (category) params.category = category;
        const response = await this.client.get('/expenses/', { params });
        return response.data;
    }

    async get(expenseId: string): Promise<Expense> {
        const response = await this.client.get(`/expenses/${expenseId}/`);
        return response.data;
    }

    async create(data: Partial<Expense>): Promise<Expense> {
        const response = await this.client.post('/expenses/', data);
        return response.data;
    }

    async update(expenseId: string, data: Partial<Expense>): Promise<Expense> {
        const response = await this.client.patch(`/expenses/${expenseId}/`, data);
        return response.data;
    }

    async delete(expenseId: string): Promise<void> {
        await this.client.delete(`/expenses/${expenseId}/`);
    }

    async listCategories(): Promise<ExpenseCategory[]> {
        const response = await this.client.get('/expense-categories/');
        return response.data;
    }
}

File: sdks/rust/src/resources/expenses.rs
use crate::models::{PaginatedResponse, Expense};
use crate::Puxbay;
use crate::PuxbayError;
use reqwest::Method;

pub struct ExpensesResource<'a> {
    pub(crate) client: &'a Puxbay,
}

impl<'a> ExpensesResource<'a> {
    pub fn new(client: &'a Puxbay) -> Self {
        Self { client }
    }

    pub async fn list(&self, page: i32, category: Option<&str>) -> Result<PaginatedResponse<Expense>, PuxbayError> {
         let mut query = format!("expenses/?page={}", page);
        if let Some(c) = category {
            query.push_str(&format!("&category={}", c));
        }

        self.client.request(
            Method::GET,
            &query,
            None::<()>,
        ).await
    }

    pub async fn get(&self, expense_id: &str) -> Result<Expense, PuxbayError> {
        self.client.request(
            Method::GET,
            &format!("expenses/{}/", expense_id),
            None::<()>,
        ).await
    }

    pub async fn create(&self, expense: &Expense) -> Result<Expense, PuxbayError> {
        self.client.request(
            Method::POST,
            "expenses/",
            Some(expense),
        ).await
    }

    pub async fn update(&self, expense_id: &str, expense: &Expense) -> Result<Expense, PuxbayError> {
        self.client.request(
            Method::PATCH,
            &format!("expenses/{}/", expense_id),
            Some(expense),
        ).await
    }

    pub async fn delete(&self, expense_id: &str) -> Result<(), PuxbayError> {
        self.client.request::<serde_json::Value>(
            Method::DELETE,
            &format!("expenses/{}/", expense_id),
            None::<()>,
        ).await?;
        Ok(())
    }

    pub async fn list_categories(&self) -> Result<Vec<String>, PuxbayError> {
        self.client.request(
            Method::GET,
            "expense-categories/",
            None::<()>,
        ).await
    }
}
